### PR TITLE
Allow custom job name in backend.run()

### PR DESCRIPTION
**Pull request by jyu00**
### Summary
This is part of #118 . It allows a user to specify a job name when calling `backend.run()` and subsequently use it as a filter to `backend.jobs()`.

### Details and comments
I ended up adding a new attribute `_job_name` to `IBMQJob` since that's the most logical place. The downside is it might make people think that attribute is always available when it's not.

### CODE REVIEW

File: qiskit/providers/ibmq/ibmqbackend.py [hunk 1]
```diff
@@ -153,7 +156,7 @@ def defaults(self, refresh=False):
 
         return self._defaults
 
-    def jobs(self, limit=10, skip=0, status=None, db_filter=None):
+    def jobs(self, limit=10, skip=0, status=None, db_filter=None, job_name=None):
```
Comment (diego-plan9) [inline qiskit/providers/ibmq/ibmqbackend.py]: Can you reorder the arguments a bit, leaving `db_filter` as the last one? It has some penalty in regards to the signature of the function being backwards-compatible, but in this case I think we gain enough clarity by keeping the function parameters `(pagination, specific small filters, most generic filter)` (specially as the specific small filters are likely to be expanded in the future).

File: qiskit/providers/ibmq/job/ibmqjob.py [hunk 2]
```diff
@@ -161,10 +165,12 @@ def __init__(self, backend, job_id, api, qobj=None,
             validate_qobj_against_schema(qobj)
             self._qobj_payload = qobj.to_dict()
             self._status = JobStatus.INITIALIZING
+            self._job_name = job_name
```
Comment (diego-plan9) [inline qiskit/providers/ibmq/job/ibmqjob.py]: Instead of setting an attribute, can you explore just passing the name to `.submit()`? As you mention, introducing a class-level attribute can be a bit confusing both for users ... and for maintainers (specially since the attribute is currently _not_ set for retrieved jobs, and the job class is quite convoluted). This hopefully will be enough for covering the requirements for this iterations, and we can include the name in a more controlled way during #104 .

File: test/ibmq/test_ibmq_job.py [hunk 3]
```diff
@@ -433,6 +433,37 @@ def test_pulse_job(self, qe_token, qe_url):
...
+        # Use a unique job name
+        job_name = str(time.time()).replace('.', '')
+        job_ids = []
+        for _ in range(2):
+            job_ids.insert(0, backend.run(qobj, job_name=job_name).job_id())
+
+        job_list = backend.jobs(job_name=job_name)
+        self.assertEqual(len(job_list), 2)
+        job_list_ids = [job.job_id() for job in job_list]
+        self.assertEqual(job_ids, job_list_ids)
```
Comment (diego-plan9) [inline test/ibmq/test_ibmq_job.py]: Can you tweak the test for comparing using sets instead of list (or a similar approach), to prevent failing in some edge cases?

File: qiskit/providers/ibmq/api_v2/clients/account.py [hunk 4]
```diff
@@ -123,30 +123,34 @@ def list_jobs_statuses(
         return self.client_api.jobs(limit=limit, skip=skip,
                                     extra_filter=extra_filter)
 
-    def job_submit(self, backend_name: str, qobj_dict: Dict[str, Any]) -> Dict[str, Any]:
+    def job_submit(self, backend_name: str, qobj_dict: Dict[str, Any],
+                   job_name: str = None) -> Dict[str, Any]:
```
Comment (diego-plan9) [inline qiskit/providers/ibmq/api_v2/clients/account.py]: Pinging @luisg5 for recommendations on the type hints - `Optional`?

**Comment by yeralin** [comment]
What happens if you pass an empty string as a `job_name`?

**Comment by jyu00** [comment]
> What happens if you pass an empty string as a `job_name`?

It will be ignored per
```
        if job_name:
            api_filter['name'] = job_name
```

**Comment by yeralin** [comment]
I guess what I was wondering is what are the constraints on the `job_name` parameter? (length, character encoding, allows special characters, etc.) 

I also could not find it on the API side. The API team hasn't implemented the feature yet?

**Comment by diego-plan9** [comment]
Thanks @jyu00 ! During the merge at e0e50813020197809ef79f2cbea5d216ef07de7f, I had to solve some conflicts in the signatures which were a result of the merging of #299 - which seem to have gone nicely!

> I guess what I was wondering is what are the constraints on the job_name parameter? (max/min length, character encoding, allows special characters, etc.)

Good question - albeit something that can be answered at the API side: an error will be returned during the call if the parameters (including the job name) do not fulfill the constraints set by the API. If it ends up being too generic, we can revise and add some informative handling, but for the time being let's just rely on the API.

File: qiskit/providers/ibmq/api_v2/clients/account.py (hunk 4 above)
Comment (luisg5) [reply qiskit/providers/ibmq/api_v2/clients/account.py]: Thanks Diego! Yes, to be consistent with other type hints already defined, it might be best to specify the type hint for `job_name` as an `Optional[str]`. This would go for other function declarations as well. Essentially, it means the same thing, but abides to PEP 484, which we are trying to follow.

**Comment by aasfaw** [comment]
@diego-plan9 I'm working on the latest master (see version below) and getting an error when I run `execute` without passing a `job_name`. Is the plan to make it mandatory to pass an empty string when `job_name` is unneeded? For the record, it works fine when I use a simulator as the backend, but breaks when I use a real quantum system as the backend. So seems like IBMQjob is the issue.

Code:
```
job = execute(circuit, backend = quantumsystem, shots=1024)
```

Version:
```
{‘qiskit-terra’: ‘0.9.0.dev0+b5d1511’,
‘qiskit-ignis’: ‘0.2.0.dev0+d5e7c7b’,
‘qiskit-aqua’: ‘0.5.6.dev0+01d5b1a’,
‘qiskit-ibmq-provider’: ‘dev-93fdcdb246a2b4ff4715a11f027c048b772e38ca’,
‘qiskit’: None,
‘qiskit-aer’: ‘0.2.3’}
```

Error:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-23-3909d03ec104> in <module>
----> 1 job = execute(circuit, backend = quantumsystem, shots=1024)
      2 from qiskit.tools.monitor import job_monitor
      3 job_monitor(job)

~/qiskit-terra/qiskit/execute.py in execute(experiments, backend, basis_gates, coupling_map, backend_properties, initial_layout, seed_transpiler, optimization_level, pass_manager, qobj_id, qobj_header, shots, memory, max_credits, seed_simulator, default_qubit_los, default_meas_los, schedule_los, meas_level, meas_return, memory_slots, memory_slot_size, rep_time, parameter_binds, **run_config)
    220 
    221     # executing the circuits on the backend and returning the job
--> 222     return backend.run(qobj, **run_config)

TypeError: run() missing 1 required positional argument: 'job_name'
```

**Comment by jyu00** [comment]
@aasfaw Thank you for catching this! `job_name` is meant to be an optional parameter and was made required incorrectly during code merge. It will be fixed by #309.
